Add decoding and encoding to string for Graph

tehPlayer opened the pull request:
I wanted to add on-fly conversion, without saving files anywhere on disk, but send the stream directly to GCloud. As I know it is possible using the low-level API, but it was missing from high-level API.
This PR shouldn't affect current I/O, it just adds new features.

naps62 commented on lib/imageflow/graph_runner.ex:
```diff
@@ -7,8 +7,8 @@ defmodule Imageflow.GraphRunner do
-         :ok <- Native.destroy(job) do
-      :ok
+         {:ok, results} <- print_results(job, graph.outputs) do
+      if results == [], do: :ok, else: {:ok, results}
```
I'm fine with a breaking change here, if a new result type needs to be used.

Returning different types based on the result leads to an inconsistent API.
I'd rather return `{:ok, []}`, or even `{:ok, %Result{}}` (with a result object that we could later mutate further if needed)`.

naps62 replied:
Another option would be to just add a separate `Graph.get_results(job)` function to retrieve these values afterwards.

Might be worth checking how the .NET wrapper does it, since this one was mostly based on it

tehPlayer replied:
I actually liked the `get_results` idea, but I still had to change return structure.

naps62 commented on lib/imageflow/graph.ex:
```diff
@@ -120,6 +132,37 @@ defmodule Imageflow.Graph do
+    graph
+    |> add_output(io_id, :bytes)
+    |> append_node(%{encode: %{io_id: io_id, preset: preset_for(encoder, opts)}})
+  end
```
it seems this is mostly duplicate code from `encode_to_file/4`, correct? How do you feel about refactoring to extract the duplication to a separate function?

naps62 commented on lib/imageflow/graph_runner.ex:
```diff
@@ -29,10 +30,13 @@ defmodule Imageflow.GraphRunner do
+      {id, value}, :ok ->
+        case value do
+          {:file, _path} -> Native.add_output_buffer(job, id)
+          :bytes -> Native.add_output_buffer(job, id)
```
I'm not understanding what the need for this `case` is. In both scenarios, you're doing the exact same thing. Why isn't the existing `with` enough?

tehPlayer replied:
My bad, I was trying something and forgot to switch it back.

naps62 reviewed (commented):
@tehPlayer thanks for the contribution! the overall idea looks good, it's something I postponed because I didn't need it on a first version, glad to have help.

I left some comments on things I should be addresses

Also, could you add a new integration test using these new flows? It's the only way to make sure the encode/decode process is actually working
`GraphTest` only checks the arguments generated, but doesn't actually do any processing

tehPlayer commented:
I've added required fixes and added an integration test. I've also changed the workflow action, because `actions/setup-elixir@v1` is no longer maintained.

naps62 commented on lib/imageflow/graph_runner.ex:
```diff
@@ -1,14 +1,36 @@
          :ok <- add_outputs(job, graph.outputs),
          :ok <- send_task(job, graph),
-         :ok <- save_outputs(job, graph.outputs),
-         :ok <- Native.destroy(job) do
```
It seems this step got left out entirely.
I'm not sure where to fit it now, since I'm assuming we need to do this only after `get_results`, but it doesn't seem correct to never de-allocate the job

naps62 commented on lib/imageflow/graph_runner.ex:
```diff
@@ -1,14 +1,36 @@
+    end
+  end
+
+  def get_results(job, %Graph{outputs: outputs} = graph) do
```
I know I'm the one who suggested this function in the first place, but the API doesn't look correct. As per Elixir standards, `Graph` should at least be the first argument.
But even with that, the fact that the caller needs to keep track of two separate values (the original `graph`, and the new `job`) makes this a clunky API

Also, coupling this with my other comment about `Native.destroy` being missing, wouldn't it be better to just make this a private function, and call it automatically in the original run flow?

naps62 commented on lib/imageflow/graph_runner.ex:
```diff
@@ -1,14 +1,36 @@
+
+        {:file, path} ->
+          {:cont, {:ok, path}}
+      end
```
I might be missreading this, but isnt' this resulting in an `:output` that corresponds just to whatever the last output from the job was?
I'm guessing it should instead build a map of all the existing outputs
